[Clang][C++20] Implement constexpr std::bit_cast for bit-fields

sethp opened the pull request:
After this commit, clang permits constructions like:
```c++
struct bits {
    unsigned char : 7;
    unsigned char flag : 1;
}

static_assert(std::bit_cast<bits>(0x80).flag); // succeeds on little-endian systems
```

A few notes for the reviewers:
- ~It's "stacked" on top of a change making `static_assert` print more complex expressions and binary operator call operands because that makes it a whole lot easier to debug, and it seems like there's a lot of overlap between folks working in both areas. I tried to open an actual "stacked" pair of PRs per the latest on #56636 , but I don't have the prerequisite permission to push to `users/sethp/...` and I'm not quite sure how to acquire it; if this doesn't work for you, just let me know what you'd prefer.~ Split out into #74852
- Overall, I feel pretty good about how the point-of-use side worked out (i.e. in `constexpr-builtin-bit-cast.cpp`); I'm especially happy that I worked out how to get the compiler to indicate which bits were indeterminate, because that really helped me learn and understand the way the structures were being laid out in memory (especially when paired with the "stacked" change).
- The implementation side is definitely a bit messier. I'm intending to continue "factoring around" the bit-field allocation logic (as mentioned in D62825), but I'm definitely not an expert in either LLVM or C++, so suggestions would be very welcome. ~~I also have an inkling that `APInt` is a better spelling of "fixed-width bit-buffer" than `SmallVector`, especially for this situation where we're (so far) exclusively mapping to/from `APInt`s.~~ I think this part at least worked out well in [2e0d3f8](https://github.com/llvm/llvm-project/pull/74775/commits/2e0d3f81e4f5b623bb476bfac0278cfc6d1bd4bc)
- As mentioned, I'm also relatively new to C++ (well, modern C++, used in anger; I don't think cutting my teeth on the language back when C++98 was the shiny new thing really helps). That said, I've got a fairly robust background in programming and computer architecture going back some 15 years. I mention that mostly because this seems like an area with more than a few subtleties, and I'm eager for your feedback on them, even if I'm not yet fluent in the C++ dialect. I did go through the [LLVM Coding Standards](https://llvm.org/docs/CodingStandards.html) and did my best to observe its recommendations; while that document was very helpful, I know it's not likely to be complete!

To better inform the shape of work that I see as outstanding here, I was hoping to solicit your feedback on three specific points:
1. Sign extension; it seems the way the constant evaluator represents bit-fields is to use a full-width `APSInt`. That's convenient for a couple reasons (e.g. there's no need for "implicit" up/down-cast nodes in the expression tree), but it does mean that we've got to invent values for some bits that really aren't "part" of the bit-field. The least surprising way I found to do this was via sign-extension, but  that does lose track of information: for example, given `struct { signed : 1 ; signed : 31  } S`, `std::bit_cast<signed>(S)` doesn't work in a `constexpr` context, but `std::bit_cast<signed>(Ex.signed)` does. I'm not really sure what there is to be done about this one besides changing the evaluator to be bit-width-precise, which seems like a larger unit of work. It's also not really new to this change, but it's something that I noticed that seemed worth raising.
2. Error message(s); I like that the "bad bits" note indicates what ranges the evaluator was looking for and didn't find, but I'm hoping you can tell me how clearly it seems to present that. There's also the classic "first-failure recursive abort" problem that can make compilers feel so nit-picky; right now, "bad bits" only notifies about the first field the evaluator sees that it can't satisfy with a value, even if there are more such objects in the cast expression. It's also a bummer that it doesn't work for the "valid uninitialized holders" (`unsigned char`, `std::byte`). I think I can fix the former by computing the entire "mask" in one pass and then applying it in the next, so we can do a total validity check once (instead of once per field). I'm a little stumped about the latter, though, since it seems like it'd mostly just be false positives: is it possible to add some sort of "deferred" note that doesn't get emitted unless they try to use the uninitialized bytes somehow?
3. Future (type) completeness; Having the benefit of "beginners mind" here, I spent a bunch of time convincing myself that the evaluator was complete in terms of the types it currently handles. I believe it is, but if a future standard invents a new type of bit-field or a new, non-BuiltinType category of TriviallyCopyable bits, it looks to me like it'd be relatively easy to omit support for those in the bit_cast implementation. I wondered "How could we bring 'bit_cast' to the attention of an implementer working elsewhere in clang?", and took a wild swing at this in the test file. But tests are only one way to communicate this kind of thing, and I suppose the constant evaluator has plenty of other possible mismatch points between it and codegen. It seems to me that attempting to match 100% of the details by construction isn't possible, so I guess I'm curious what other strategies y'all have employed here?

Thanks for all the hard work y'all do!

sethp committed:
[Clang][Sema] Print more static_assert exprs

This change introspects more values involved in a static_assert, and
extends the supported set of operators for introspection to include
binary operator method calls.

It's intended to address the use-case where a small static_assert helper
looks something like this (via `constexpr-builtin-bit-cast.cpp`):

```c++
struct int_splicer {
  unsigned x;
  unsigned y;

  constexpr bool operator==(const int_splicer &other) const {
    return other.x == x && other.y == y;
  }
};
```

When used like so:

```c++
constexpr int_splicer got{1, 2};
constexpr int_splicer want{3, 4};
static_assert(got == want);
```

Then we'd expect to get the error:

```
Static assertion failed due to requirement 'got == want'
```

And this change adds the helpful note:

```
Expression evaluates to '{1, 2} == {3, 4}'
```

sethp committed:
[Clang][C++20] Implement constexpr std::bit_cast for bit-fields

After this commit, clang permits constructions like:
```c++
struct bits {
    unsigned char : 7;
    unsigned char flag : 1;
}

static_assert(std::bit_cast<bits>(0x80).flag); // succeeds on little-endian systems
```

This change builds on the prior work in https://reviews.llvm.org/D62825

sethp commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -49,19 +48,33 @@
 #include <functional>
+#include <iomanip>
+#include <iterator>
 #include <optional>
```
Not entirely sure how I'm getting away with this on Linux/gcc, but it looks like the CI build failed because of a missing #include:

```suggestion
#include <optional>
#include <sstream>
```

tbaederr commented:
I don't have the capacity to review this properly, but the changes to the `static_assert` diagnostics should be split out IMO.

sethp commented:
Happy new year!

An update from my end: I've made what feels like good progress on refactoring the implementation to be a little more concise, both in terms of code and dynamic instruction count (especially for large types). There's still a little more I'd like to do there, especially as relates to handling host-target endianness mismatches: and I think at least one more test I need to devise to validate that the implementation handles bit-fields split across a byte boundary correctly. I hope to push that work up soon, but it still feels a little "too early" to be properly respectful of y'all's time.

However, in doing that work, I identified an issue with `bool` types (and, probably, bit-precise `_BitInt(N)` types): the work represented here doesn't yet handle them. Instead, it triggers an assertion when the evaluator produces an 8-bit result where the expression wanted 1-bit. I'm working on a fix for that, but doing so requires handling the case when trying to cram something other than a 0 or 1 into a bool, which has a few wrinkles to it. The way I understand `constexpr` is that _if_ it produces a value, then it will match the behavior specified at runtime, but the evaluator can make the principled choice to be incomplete (as previously with bit-casting bit-fields)—is that right? So, if I'm finding cases where the bit-cast wouldn't match the runtime behavior, it'd be possible to refuse to produce a value by producing an error instead?

sethp committed:
refactor: BitCastBuffer with APInt, fix for bools

Make use of the pre-exising APInt type as a pre-established spelling of
BitVector to get much more granular with the BitCastBuffer; this also
permits more thorough usage of `const APInt&` references to make the
copies easier to optimize away and reduces the common case of
int->buffer->int from 4-7 copies down to 2. As a bonus, we also got a
bit more flexible with supporting platforms where `CHAR_BIT` is
something other than 8, addressing some of the review feedback from the
original diff, though there's plenty of follow-up work here and
elsewhere to make that a reality.

Additional, uncovers and fixes some errant behavior around `bool`s
introduced in the previous commit, and makes the handling of
`_BitInt(N)` types more consistent while preserving
backwards-compatability.

sethp commented on clang/test/SemaCXX/constexpr-builtin-bit-cast.cpp:
```diff
@@ -404,29 +691,126 @@ constexpr unsigned char identity3b = __builtin_bit_cast(unsigned char, identity3
+// expected-note@+1 {{cannot be represented in type 'bool'}}
+constexpr bool test_bad_bool = __builtin_bit_cast(bool, 'A'); // expected-error {{must be initialized by a constant expression}}
+
+static_assert(round_trip<signed char>(true));
+static_assert(round_trip<unsigned char>(true));
```
These tests document the existing behavior, but both this and the `bit_cast<uint8_t>(false)` below are somewhat troublesome in that we're promoting the `bool`'s padding bits to "real" bits (by zero-extension). As I understand it, that's allowed by the spec (padding bits are "unspecified"), but it does mean that we've got to "over-read" `bool`s as being a full `i8` when we convert `bit_cast<bool>((uint8_t)N)` so that we can refuse to produce a constant value if the top bits are non-zero.

It would be more consistent and simplify the implementation a bit to throw an error on `bit_cast<char>(false)` in a constexpr context complaining about attempting to read padding bits, but I've chosen to preserve the existing behavior here in service of backwards compatibility and usability.

sethp commented on clang/test/SemaCXX/constexpr-builtin-bit-cast-bitint.cpp:
```diff
@@ -0,0 +1,82 @@
+namespace bit_precise {
+// ok so it's a little bit of a lie to say we don't support _BitInt in any casts; we do in fact
+// support casting _from_ a _BitInt(N), at least some of the time
+static_assert(bit_cast<uint8_t, _BitInt(8)>(0xff) == 0xff);
+template <int N> struct bytes { uint8_t b[N]; };
+static_assert(bit_cast<bytes<2>, _BitInt(12)>(0xff).b[(LITTLE_END ? 0 : /* fixme */ 0)] == 0xff);
+static_assert(bit_cast<bytes<4>, _BitInt(24)>(0xff).b[(LITTLE_END ? 0 : /* fixme */ 2)] == 0xff);
```
These tests *also* document the existing (buggy) behavior; I don't expect there's a lot of people out there using constexpr `std::bit_cast` on `_BitInt(N)` types, but if there are they're having a bad time with it.

That said, I took pains to preserve the existing behavior to try and make this change purely "additive"—to the best of my ability I tried to preserve the property that no code that previously compiled without error (including badly formed constructs such as these) would start producing an error. 

That's a pretty strong stance though, that I went with because I couldn't find details on `clang`'s backwards compatibility policy. I'm happy to soften it: in this case, I would probably prefer to produce an error instead of a value from `APValueToBufferConverter` when it encounders a `_BitInt(N)` type and thereby make it consistently unsupported pending future work.

sethp commented on clang/test/SemaCXX/constexpr-builtin-bit-cast.cpp:
```diff
@@ -404,29 +691,126 @@ constexpr unsigned char identity3b = __builtin_bit_cast(unsigned char, identity3
+constexpr auto _bad_bool4 = __builtin_bit_cast(pad, s);
+
+
+// `bool` includes padding bits, but *which* single bit stores the
+// value is under-specified. These tests not-so-secretly assert that
+// it's in fact the LSB that the compiler "sees" as the value.
+struct pack {
+  bool a : 1;
+  bool b : 1;
+
+  // 1 bit of value, 5 bits of padding
+  bool c : 6;
+};
+
+constexpr auto packed = bit_cast<pack, uint8_t>(LITTLE_END ? 0x07 : 0xc1);
+static_assert(packed.a && packed.b && packed.c);
```
Taken together this makes a very strong assertion about the ABI of a `boolean`; that seems to match the existing behavior, both in how the evaulator handled 8-bit values and the output from codegen: https://godbolt.org/z/xsjE7aG3e

Probably there's no good reason to make a different decision in the future, but it does seem worth calling out: this is far from the only place the evaluator rests on the decisions made elsewhere in clang/llvm-land, but it's (partially) a new one.

sethp commented on clang/test/SemaCXX/constexpr-builtin-bit-cast.cpp:
```diff
@@ -485,16 +869,16 @@ typedef bool bool17 __attribute__((ext_vector_type(17)));
+static_assert(check_round_trip<bool32>(static_cast<int>(0xCAFEBABE)));
+static_assert(check_round_trip<bool128>(static_cast<__int128_t>(0xCAFEBABE0C05FEFEULL)));
 
 // expected-error@+2 {{constexpr variable 'bad_bool9_to_short' must be initialized by a constant expression}}
 // expected-note@+1 {{bit_cast involving type 'bool __attribute__((ext_vector_type(9)))' (vector of 9 'bool' values) is not allowed in a constant expression; element size 1 * element count 9 is not a multiple of the byte size 8}}
```
I don't know if it's important or not, but this would be pretty easy to support with the new bit-wise `APInt`-backed BitCastBuffer: I didn't try for it because there's already quite a lot going on here, but I'm happy to follow up on that if it seems worthwhile.

sethp commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -6901,51 +6916,260 @@ bool HandleOperatorDeleteCall(EvalInfo &Info, const CallExpr *E) {
+  /// The offset of the bitfield storage from the start of the struct.
+  const CharUnits StorageOffset;
+
+  static BitFieldInfo MakeInfo(const ASTContext &Ctx, const FieldDecl *FD,
+                               const ASTRecordLayout &Layout) {
```
This is part-way to the aforementioned refactoring of the bit-field layout & access logic between AST and codegen; this `MakeInfo` method happens to make the same choices that codegen does, but those choices seem very consistent across time and even compilers (sorry, I lost the godbolt link for this one).

Really it's only two:

1. What happens with "oversized" bit-fields, and
2. What "direction" does the offset count on little-endian vs. big-endian?

At this point, the answers seem consistent; most of the complicated stuff already got handled in setting up the bit-wise field offsets in the first place. 

That said, at this point I'm a little stuck: I can move logic around between here, CGRecordBuilder/CGRecordLowering, and `AST/RecordLayoutBuilder.cpp`, but I don't have a sense of clarity around the outcome I'd be looking for there. I suspect the goal is shaped something like "someone looking to answer those two questions differently affects the constant evaluator with their changes", but I'm not sure how I'd achieve that.

sethp commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -6901,51 +6916,260 @@ bool HandleOperatorDeleteCall(EvalInfo &Info, const CallExpr *E) {
+
+  // copyBitsFrom acts like `LHS[Dst] = RHS[Src];`, if `APInt`s supported
+  // slicing
+  inline static void copyBitsFrom(APInt &LHS, const BitSlice &Dst,
+                                  const APInt &RHS, const BitSlice &Src) {
```
This has roughly the right API shape to it, I feel, but it's a bit of a mismatch between this and the existing APInt `insertBits` method, especially since the latter contains this snippet:

```
  // General case - set/clear individual bits in dst based on src.
  // TODO - there is scope for optimization here, but at the moment this code
  // path is barely used so prefer readability over performance.
  for (unsigned i = 0; i != subBitWidth; ++i)
    setBitVal(bitPosition + i, subBits[i]);
```

We'd fairly often be falling through to that "general" implementation here, so I went ahead and improved matters some in https://github.com/sethp/llvm-project/pull/3 : but that's an time-domain-effect-only change, and not functionally required, so my plan is to wait for this PR to land before proposing those optimizations. I'm noting it here just in case you also happen to notice that little wrinkle.

sethp commented:
oh, @tbaederr I see you've done some overlapping work over in https://github.com/llvm/llvm-project/pull/68288 : I'll make time to review that and see what ought to happen to dovetail these patches.

frederick-vs-ja commented:
This PR should fix #54018.
